From 04149201432d7448419fadd40462bb5e1d63eeec Mon Sep 17 00:00:00 2001
From: vipinc <vipinc@ami.com>
Date: Wed, 24 Jul 2024 12:23:22 +0530
Subject: [PATCH] Add IPMI SEL Clear Log Support

This commit add support to SEL Cleared Log to IPMI.

Before Fix:
root@AMIOT-7ace89ca07d9:~# ipmitool sel list
   1 |  Pre-Init  |0000000000| reserved |  | Asserted
root@AMIOT-7ace89ca07d9:~# ipmitool sel clear
Clearing SEL.  Please allow a few seconds to erase.
root@AMIOT-7ace89ca07d9:~#
root@AMIOT-7ace89ca07d9:~# ipmitool sel list
   1 |  Pre-Init  |0000000000| reserved |  | Asserted
root@AMIOT-7ace89ca07d9:~#

After Fix:
root@AMIOT-ca696b3e90f8:~# ipmitool sel list
SEL has no entries
root@AMIOT-ca696b3e90f8:~# ipmitool sel clear
Clearing SEL.  Please allow a few seconds to erase.
root@AMIOT-ca696b3e90f8:~# ipmitool sel list
   1 | 07/24/24 | 06:43:09 UTC | Event Logging Disabled #0xff | Log area reset/cleared | Asserted

Signed-off-by: vipinc <vipinc@ami.com>
---
 log_manager.hpp | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/log_manager.hpp b/log_manager.hpp
index be76dfd..a357345 100644
--- a/log_manager.hpp
+++ b/log_manager.hpp
@@ -29,6 +29,22 @@ using Severity = sdbusplus::xyz::openbmc_project::Logging::server::Entry::Level;
 using LogsCleared =
     sdbusplus::xyz::openbmc_project::Logging::Error::LogsCleared;
 
+namespace ipmi
+{
+constexpr auto strEventDir = "EVENT_DIR";
+constexpr auto strGenerateId = "GENERATOR_ID";
+constexpr auto strRecordType = "RECORD_TYPE";
+constexpr auto strSensorData = "SENSOR_DATA";
+constexpr auto strSensorPath = "SENSOR_PATH";
+constexpr auto strSensorType = "SENSOR_TYPE";
+
+constexpr auto systemEventRecord = "2";
+constexpr auto bmcFirmware = "32";
+constexpr auto assertEvent = "1";
+constexpr auto logClearEvent = "02ffff";
+constexpr auto eventLoggingDisabled = "16";
+} // namespace ipmi
+
 namespace details
 {
 template <typename... T>
@@ -367,8 +383,17 @@ class Manager : public details::ServerObject<DeleteAllIface, CreateIface>
     {
         log<level::INFO>("Deleting all log entries");
         auto numbersOfLogs = manager.eraseAll();
+        std::map<std::string, std::string> ipmiLogClear = {
+            {ipmi::strEventDir, ipmi::assertEvent},
+            {ipmi::strRecordType, ipmi::systemEventRecord},
+            {ipmi::strGenerateId, ipmi::bmcFirmware},
+            {ipmi::strSensorData, ipmi::logClearEvent},
+            {ipmi::strSensorPath, ""},
+            {ipmi::strSensorType, ipmi::eventLoggingDisabled}};
+
         std::map<std::string, std::string> additionalData;
         additionalData.emplace("NUM_LOGS", std::to_string(numbersOfLogs));
+        additionalData.insert(ipmiLogClear.begin(), ipmiLogClear.end());
         manager.create(LogsCleared::errName, Severity::Informational,
                        additionalData);
     }
-- 
2.25.1

